core/domain/exp_domain.py

```diff
@@ -2514,12 +2515,1063 @@ def _convert_states_v50_dict_to_v51_dict(
+            default_outcome = state_dict['interaction']['default_outcome']
+            if default_outcome is not None:
+                def_ref_exp_id = default_outcome['refresher_exploration_id']
+                if def_ref_exp_id is not None:
+                    state_dict['interaction']['default_outcome'][
+                        'refresher_exploration_id'] = None
+
+            state_dict['interaction']['answer_groups'] = answer_groups
+        return states_dict
+
+    @classmethod
+    def _choices_should_be_unique_and_non_empty(
+        cls, choices: List[str],
+        answer_groups: List[state_domain.AnswerGroup]
+    ) -> Tuple(Any):
+        """Handles choices present in the ItemSelectionInput or
+        in MultipleChoiceInput interactions, implements the following -
+            - If only one choice is empty then simply removes it
+            - If multiple choices are empty replace them with `Choice 1` ,
+            `Choice 2` etc
+            - If choices are duplicate, removes the later choice
+
+        Args:
+            choices: List[str]. A list of choices.
+
+        Returns:
+            choices_to_save: List[str]. The list of valid choices.
+        """
+        # Misbehaving when after removing invalid choices and the choices count becomes below 4.
+        empty_choices = []
+        seen_choices = []
+        choices_to_remove = []
+        invalid_choices_index = []
+        for choice in choices:
+            if choice['html'].strip() in ('<p></p>', ''):
+                empty_choices.append(choice)
+
+        # Only one choice is empty.
+        if len(empty_choices) == 1:
+            invalid_choices_index.append(choices.index(empty_choices[0]))
+            choices.remove(empty_choices[0])
+        # Multiple choices are empty.
+        else:
+            for idx, empty_choice in enumerate(empty_choices):
+                empty_choice['html'] = (
+                    '<p>' + 'Choice ' + str(idx+1) + '</p>'
+                )
+
+        # Duplicate choices.
+        for choice in choices:
+            if choice['html'] not in seen_choices:
+                seen_choices.append(choice['html'])
+            else:
+                choices_to_remove.append(choice)
+                invalid_choices_index.append(choices.index(choice))
+
+        for choice_to_remove in choices_to_remove:
+            choices.remove(choice_to_remove)
+
+        # Remove rules whose choice has been deleted.
+        for answer_group in answer_groups:
+            invalid_rules = []
+            for rule_spec in answer_group['rule_specs']:
+                if rule_spec['rule_type'] == 'Equals':
+                    if rule_spec['inputs']['x'] in invalid_choices_index:
+                        invalid_rules.append(rule_spec)
+            for invalid_rule in invalid_rules:
+                answer_group['rule_specs'].remove(invalid_rule)
+            # Removal of answer group if it is valid check.
```

Comment:
What happens when the answer group ends up with 0 rule specs?

Reply:
I have added the checks in #15846  to gather the result for, if the whole answer_group is invalid and if it's destination node is not `try again`. That's why I just left a comment so that I can add the details later.